feat: add progress bars for time-consuming operations

Eshaan-byte opened the pull request:
This PR implements visual progress feedback for all time-consuming Dora CLI operations using the indicatif crate.

Changes:
- Created shared progress bar component library (progress.rs)
- Added progress to build, up, start, and run commands
- Unified cyan/blue styling with success/failure indicators
- Smart TTY detection (auto-hides in CI environments)
- Multi-progress support for parallel operations

All operations >2s now have visual feedback with consistent behavior.

Testing:
- cargo check: passed
- cargo fmt: passed
- cargo clippy: passed

Closes #1229

Eshaan-byte committed:
feat: add progress bars for time-consuming operations

Closes #1229

This PR implements visual progress feedback for all time-consuming
Dora CLI operations using the indicatif crate.

Changes:
- Created shared progress bar component library (progress.rs)
- Added progress to build, up, start, and run commands
- Unified cyan/blue styling with success/failure indicators
- Smart TTY detection (auto-hides in CI environments)
- Multi-progress support for parallel operations

All operations >2s now have visual feedback with consistent behavior.

Testing:
- cargo check: passed
- cargo fmt: passed
- cargo clippy: passed

phil-opp commented on binaries/cli/Cargo.toml:
```diff
@@ -57,6 +57,8 @@ crossterm = "0.29.0"
 itertools = "0.14"
 sysinfo = "0.36.1"
+indicatif = "0.17"
+atty = "0.2"
```
The `atty` crate is [deprecated](https://github.com/softprops/atty?tab=readme-ov-file#atty). Please use https://doc.rust-lang.org/std/io/trait.IsTerminal.html instead.

phil-opp commented on binaries/cli/src/progress.rs:
````diff
@@ -0,0 +1,433 @@
+//! // do work...
+//! pb1.finish();
+//! pb2.finish();
+//! ```
````
Please also provide an example that shows an actual progress bar, instead of just spinners.

phil-opp commented on binaries/cli/src/progress.rs:
```diff
@@ -0,0 +1,433 @@
+}
+
+/// Builder for creating progress bars with custom configuration.
+pub struct ProgressBarBuilder {
```
Is this used anywhere (except for tests)?

phil-opp commented on binaries/cli/src/command/build/local.rs:
```diff
@@ -98,24 +114,53 @@ impl BuildLogger for LocalBuildLogger {
+                pb.set_message(format!("{}: {}", self.node_id, message_str));
+            }
+        } else {
+            // Fallback to println if no progress bar
```
I'm not sure if we want to "hide" the log messages by default. It might be good to still have the output in the terminal so that you can scroll up later to look for warnings or errors.

phil-opp reviewed (commented):
Thanks for the PR!

Eshaan-byte committed:
Address PR feedback on progress bars

This commit addresses the remaining review feedback from PR #1249:

1. Replace `atty` with `std::io::IsTerminal`:
   - Removed deprecated `atty` crate dependency
   - Use `std::io::stderr().is_terminal()` for TTY detection

2. Rewrite progress bar documentation:
   - Simplified module docs to focus on behavior and usage
   - Added progress bar example (not just spinners)
   - Clarified behavior in non-TTY environments
   - Removed marketing terms

3. Add progress bar example:
   - Added example showing progress bar with known steps

4. Remove unused code:
   - Removed `should_show_progress()` function
   - Removed `ProgressBarBuilder` struct and tests

5. Refactor to reduce duplication:
   - Created `new_inner()` helper method
   - Both `new()` and `new_bytes()` now use shared implementation

6. Fix MultiProgress in try_clone():
   - Store MultiProgress reference in LocalBuildLogger
   - Use multi.add_spinner() in try_clone() to properly add to MultiProgress

7. Keep log messages visible:
   - Always print log messages to stdout for scrollback
   - Also update progress bar message for live feedback
   - Users can now scroll up to see warnings/errors

phil-opp reviewed (commented):
Thanks for the updates!

Eshaan-byte commented on binaries/cli/src/command/build/local.rs:
```diff
@@ -137,30 +154,20 @@ impl BuildLogger for LocalBuildLogger {
+            Some(multi.add_spinner(format!("Building {}", self.node_id)))
+        } else {
+            None
+        };
```
Good catch! You're absolutely right about both issues:
"Now cloning a non-multi progress bar always results in None, doesn't it?" - Yes, the previous code would lose the progress bar if multi was None but progress_bar was Some.
"And why is it always a spinner?" - You're right, I was hardcoding it as a spinner instead of preserving the original progress bar's message.
To answer your question: Yes, try_clone() is used. It's called at [libraries/core/src/build/mod.rs:125](https://github.com/dora-rs/dora/blob/main/libraries/core/src/build/mod.rs#L125) when spawning a tokio task to log stdout from build commands. I've fixed the implementation to properly handle both cases:
If we have a MultiProgress, it adds the cloned progress bar to it (preserving the original message)
If we don't have a MultiProgress but do have a progress bar, it creates a standalone spinner (also preserving the message)
Otherwise, it returns None
The fix now uses pb.inner().message() to preserve the original progress bar's message instead of hardcoding a new one. Let me know if this looks good or if you'd prefer a different approach!

Eshaan-byte committed:
Fix try_clone() to properly preserve progress bars

Addresses review feedback about try_clone() implementation:

1. Fixed bug where cloning a progress bar without MultiProgress would
   result in None instead of preserving the progress bar

2. Now preserves the original progress bar's message instead of
   hardcoding a new message format

3. Handles three cases properly:
   - With MultiProgress: adds new spinner to multi with same message
   - Without MultiProgress: creates standalone spinner with same message
   - No progress bar: returns None

The try_clone() method is used at libraries/core/src/build/mod.rs:125
when spawning tokio tasks to log stdout from build commands.

Eshaan-byte committed:
Use Arc<ProgressBar> for cheap cloning without duplicating display

This addresses the PR feedback by wrapping ProgressBar in Arc instead
of creating new progress bars in try_clone(). This approach:

- Avoids displaying duplicate progress bars when cloning
- Provides cheap clones through Arc's reference counting
- Simplifies the try_clone() implementation significantly

The previous implementation created new progress bars on each clone,
which would result in multiple visual indicators for the same task.

phil-opp reviewed (commented):
Thanks for the updates!

The code looks good to me now, but the output format still needs improvement imo.

For me, the `dora build` output looks like this currently:

<img width="1102" height="264" alt="Image" src="https://github.com/user-attachments/assets/88c66996-a759-412f-ad81-5f1085dda88d" />

So the percentage is printed a lot and makes the output quite messy. I would prefer if we could "pin" the progress bar and the spinners to the bottom so that they aren't part of the messages that scroll up. So basically a region at the bottom where we display progress and another region above where we print the log/stdout messages.

For example, here is how `cargo` does it:

<img width="610" height="373" alt="Image" src="https://github.com/user-attachments/assets/3d1a322d-1c30-48ad-b4bf-1dd198c59bd7" />

The `Building` progress marker at the bottom stays there and does not interfere with the "Compiling" lines above.

Eshaan-byte committed:
Use MultiProgress::println to keep progress bars pinned at bottom

This addresses phil-opp's feedback about output readability by using
indicatif's MultiProgress::println instead of regular println!.

Changes:
- Added MultiProgress::println method with fallback for non-TTY
- Updated log_message to use MultiProgress::println
- Progress bars now stay pinned at bottom while logs scroll above
- Non-terminal outputs (CI/logs) still work via println! fallback

This matches cargo's behavior where progress indicators stay at the
bottom and don't interfere with scrolling log messages.

Deepak-negi11 commented:
@phil-opp
<img width="436" height="181" alt="Screenshot 2026-01-11 211748" src="https://github.com/user-attachments/assets/338fae5b-f6a3-4680-92ac-6c821336e591" />
I see this at the beginning if this is not a problem i think the pr is good to go . You can check it again if i have miss somthing

phil-opp committed:
Fix: Set progress bar messages only after adding them to `MultiProgress`

Setting the message before adding it to `MultiProgress` draws the progress bar directly, which messes up the rendering of the multi progress bar. This leads to duplicated progress bars in the output.

phil-opp reviewed (approved):
Thanks for the updates!

phil-opp reviewed (changes requested):
I noticed that we still see a lot of spinners in the logs for `dora run`:

<img width="543" height="147" alt="Image" src="https://github.com/user-attachments/assets/0b976c48-0508-4017-9449-b8ae0ab1193f" />

This needs to be fixed too before we can merge.

heyong4725 commented:
Closing — the work itself is solid, but the underlying motivation is weaker than it looks on inspection.

## What I evaluated

I checked the four commands this PR adds spinners to (`dora build`, `dora up`, `dora start`, `dora run`) against the actual user experience they produce today:

| Command | Typical duration | Already has output? |
|---|---|---|
| `dora build` | seconds to minutes | **Yes — cargo prints per-crate compile status** |
| `dora up` | <5 seconds (spawn coord+daemon, wait for ready) | minimal |
| `dora start` | <2 seconds (send command, wait ack) | minimal |
| `dora run` | varies; can be slow if dataflow has `build:` directives | mix (cargo + then logs) |

The genuinely slow operations (`dora build`, `dora run --build`) already have cargo's compile progress output. Layering a `dora`-level spinner that says "Building nodes..." on top of cargo's 200-line per-crate output creates two competing UX layers. The fast operations (`up`, `start`) are over before a typical user notices.

## The real user pain

The pain that "did `dora up` hang?" reflects is real, but it's a much smaller fix: a one-line `println!` saying "waiting for coordinator..." during the spawn/connect phase would cover 80% of it. The PR's 560-LOC framework + a new `indicatif` dependency is solving a smaller problem with a bigger hammer than it needs.

## Additional concerns

- **`dora start --attach`** streams logs from the dataflow. Interleaving spinners with log lines is the kind of UX bug that's hard to get right (cursor positioning, terminal resize, log lines longer than the terminal width). `MultiProgress::println` helps but does not fully solve it; needs careful visual testing.
- **Rebase cost is non-trivial** if this were to merge: PR is behind 5 commits including #1797 (auth flag on `start_coordinator`), the env-var-driven `coordinator_addr` in `up()`, and a `connect_with_retry` helper that replaces the manual sleep loop. The progress-bar placement was designed for the older flow.

## What would change this decision

If you've seen real users report "is dora hanging?" — especially during `dora up` or the wait-for-daemon phase of `dora run` — that'd be a different conversation. A targeted, small-scope fix could ship the perceived-latency improvement without the framework.

A fresh, small PR doing just that ("Add a `println!` in `dora up` while waiting for coordinator/daemon to connect") would be welcome.

The progress.rs library itself is well-designed (TTY auto-hide, `MultiProgress::println` fallback, tests included). If a future need arises for a richer progress UX, this code is a good reference point.

Thanks @Eshaan-byte for the careful implementation work.
